src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,411 @@
+    for (uint16_t endpoint_idx = 0; endpoint_idx < lastEndpointIndex; endpoint_idx++)
+    {
+        if (emberAfEndpointIndexIsEnabled(endpoint_idx))
+        {
+            return emberAfEndpointFromIndex(endpoint_idx);
+        }
+    }
+
+    // No enabled endpoint found. Give up
+    return kInvalidEndpointId;
+}
+
+std::optional<unsigned> CodegenDataModel::TryFindEndpointIndex(chip::EndpointId id) const
+{
+    const uint16_t lastEndpointIndex = emberAfEndpointCount();
+
+    if ((mEndpointIterationHint < lastEndpointIndex) && emberAfEndpointIndexIsEnabled(mEndpointIterationHint) &&
+        (id == emberAfEndpointFromIndex(mEndpointIterationHint)))
+    {
+        return std::make_optional(mEndpointIterationHint);
+    }
+
+    // Linear search, this may be slow
+    for (uint16_t endpoint_idx = 0; endpoint_idx < lastEndpointIndex; endpoint_idx++)
+    {
+        if (!emberAfEndpointIndexIsEnabled(endpoint_idx))
+        {
+            continue;
+        }
+
+        if (id == emberAfEndpointFromIndex(endpoint_idx))
+        {
+            return std::make_optional(endpoint_idx);
+        }
+    }
+
+    return std::nullopt;
+}
+
+EndpointId CodegenDataModel::NextEndpoint(EndpointId before)
+{
+    const unsigned lastEndpointIndex = emberAfEndpointCount();
+
+    std::optional<unsigned> before_idx = TryFindEndpointIndex(before);
+    if (!before_idx.has_value())
+    {
+        return kInvalidEndpointId;
+    }
+
+    // find the first enabled index
+    for (uint16_t endpoint_idx = static_cast<uint16_t>(*before_idx + 1); endpoint_idx < lastEndpointIndex; endpoint_idx++)
+    {
+        if (emberAfEndpointIndexIsEnabled(endpoint_idx))
+        {
+            mEndpointIterationHint = endpoint_idx;
+            return emberAfEndpointFromIndex(endpoint_idx);
+        }
+    }
+
+    // No enabled enpoint after "before" was found, give up
+    return kInvalidEndpointId;
+}
+
+InteractionModel::ClusterEntry CodegenDataModel::FirstCluster(EndpointId endpointId)
+{
+    const EmberAfEndpointType * endpoint = emberAfFindEndpointType(endpointId);
+    VerifyOrReturnValue(endpoint != nullptr, InteractionModel::ClusterEntry::Invalid());
+    VerifyOrReturnValue(endpoint->clusterCount > 0, InteractionModel::ClusterEntry::Invalid());
+    VerifyOrReturnValue(endpoint->cluster != nullptr, InteractionModel::ClusterEntry::Invalid());
+
+    return FirstServerClusterEntry(endpointId, endpoint, 0, mClusterIterationHint);
+}
+
+std::optional<unsigned> CodegenDataModel::TryFindServerClusterIndex(const EmberAfEndpointType * endpoint, chip::ClusterId id) const
+{
+    const unsigned clusterCount = endpoint->clusterCount;
+
+    if (mClusterIterationHint < clusterCount)
+    {
+        const EmberAfCluster & cluster = endpoint->cluster[mClusterIterationHint];
+        if (IsServerMask(cluster.mask) && (cluster.clusterId == id))
+        {
+            return std::make_optional(mClusterIterationHint);
+        }
+    }
+
+    // linear search, this may be slow
```

Comment:
I assume the reason this does not use `emberAfClusterIndex` is to avoid re-doing the "get the endpoint type" iteration?  Worth documenting that.